// forta-agent/dist/sdk/transaction.event.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionEvent = void 0;
var ethers_1 = require("ethers");
var lodash_1 = __importDefault(require("lodash"));
var utils_1 = require("./utils");
var TransactionEvent = (function () {
    function TransactionEvent(type, network, transaction, traces, addresses, block, logs, contractAddress) {
        if (traces === void 0) { traces = []; }
        if (logs === void 0) { logs = []; }
        this.type = type;
        this.network = network;
        this.transaction = transaction;
        this.traces = traces;
        this.addresses = addresses;
        this.block = block;
        this.logs = logs;
        this.contractAddress = contractAddress;
    }
    Object.defineProperty(TransactionEvent.prototype, "hash", {
        get: function () {
            return this.transaction.hash;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "from", {
        get: function () {
            return this.transaction.from;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "to", {
        get: function () {
            return this.transaction.to;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "gasPrice", {
        get: function () {
            return this.transaction.gasPrice;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "timestamp", {
        get: function () {
            return this.block.timestamp;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "blockNumber", {
        get: function () {
            return this.block.number;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TransactionEvent.prototype, "blockHash", {
        get: function () {
            return this.block.hash;
        },
        enumerable: false,
        configurable: true
    });
    TransactionEvent.prototype.filterLog = function (eventAbi, contractAddress) {
        return (0, utils_1.filterLog)(this.logs, eventAbi, contractAddress);
    };
    TransactionEvent.prototype.filterFunction = function (functionAbi, contractAddress) {
        var _a;
        functionAbi = lodash_1.default.isArray(functionAbi) ? functionAbi : [functionAbi];
        var sources = [
            this.transaction,
        ];
        if (this.traces.length) {
            sources = this.traces.map(function (_a) {
                var action = _a.action;
                return ({
                    data: action.input,
                    value: action.value,
                    to: action.to,
                });
            });
        }
        if (contractAddress) {
            contractAddress = lodash_1.default.isArray(contractAddress)
                ? contractAddress
                : [contractAddress];
            var contractAddressMap_1 = {};
            contractAddress.forEach(function (address) {
                contractAddressMap_1[address.toLowerCase()] = true;
            });
            sources = sources.filter(function (source) { return source.to && contractAddressMap_1[source.to.toLowerCase()]; });
        }
        var results = [];
        var iface = new ethers_1.ethers.utils.Interface(functionAbi);
        for (var _i = 0, sources_1 = sources; _i < sources_1.length; _i++) {
            var source = sources_1[_i];
            try {
                var parsedTransaction = iface.parseTransaction(source);
                results.push(Object.assign(parsedTransaction, {
                    address: (_a = source.to) === null || _a === void 0 ? void 0 : _a.toLowerCase(),
                }));
            }
            catch (e) { }
        }
        return results;
    };
    return TransactionEvent;
}());
exports.TransactionEvent = TransactionEvent;
